orderBook.js

```diff
@@ -1,5 +1,50 @@
+
 function reconcileOrder(existingBook, incomingOrder) {
-  throw new Error('Not Implemented Yet')
+  if (existingBook.length === 0) {
+    existingBook.push(incomingOrder)
+  } else {
+    var matchingIndex = -1
+    var quantityNotFulfilled = incomingOrder.quantity
+    var matchingIndexList = []
+    for (var i = 0; i < existingBook.length; i++) {
+      var existingOrder = existingBook[i]
+      if (incomingOrder.type != existingOrder.type
+        && incomingOrder.price === existingOrder.price) {
+        matchingIndex = i
+        matchingIndexList.push(i)
+        if (quantityNotFulfilled > existingOrder.quantity) {
+          quantityNotFulfilled = quantityNotFulfilled - existingOrder.quantity
+        } else {
+          quantityNotFulfilled = 0
+          break
+        }
+      }
+    }
+    var remainingQuantity = incomingOrder.quantity
+    if (matchingIndex != -1) {
+      for (var j = 0; j < matchingIndexList.length; j++) {
```

Comment:
With this many indexes and the this many references to them I would have chosen a more meaningful name than `j` as it makes it easier to track as you go through the code.